### PR TITLE
#162414162 Reset Password via email

### PR DESCRIPTION
**Pull request by Maxfurry**
#### What does this PR do?
Sends Reset Link to user email

#### Description of Task to be completed?
- Send reset link to user email
- Confirm the reset link gotten from user email

#### How should this be manually tested?
- Register on sendgrid and get API key form them
- Add API key to environment variable using variable name SENDGRID_API_KEY
- Clone this branch 
- Run npm install
- Send post request in the format
	{
	  "email":"jeosmith@myserver.com",
	}
- If successful you should see a reply like the below
{
    "success": true,
    "message": "Reset password mail sent successfully"
}
- If not successful you should see a reply like the below
{
    "success": false,
    "message": "Reset password mail not sent"
}

- If successful go to your email (if not in inbox check spam)
- Click the reset link seen
- You will get a response telling you if the link is:
	- Valid
	- Invalid
	- Expired

#### What are the relevant pivotal tracker stories?
PT Link: https://www.pivotaltracker.com/story/show/162414162

#### Screenshots
screen shot 2019-01-18 at 10 39 10 am

### CODE REVIEW

File: controllers/forgetPasswordController.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+ */
+class PasswordReset {
+/**
+ * sends reset E-mail.
+ * @param {object} req Request Object.
+ * @param {object} res Response Object.
+ * @returns {object} Response with status, success true or false, and message.
+ */
+  static async sendResetLink(req, res) {
+    // Checks if email exist
```
Comment (codrex) [inline controllers/forgetPasswordController.js]: remove inline comments

File: controllers/forgetPasswordController.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    const user = await User.getUserDetails('email', req.body.email);
+    if (!user) {
+      return res.status(404).json({
```
Comment (codrex) [inline controllers/forgetPasswordController.js]: Move constants and string literal into constants

File: lib/modelManagers/usermodel.js [hunk 3]
```diff
@@ -59,6 +59,26 @@ class User {
...
+   * @memberof User
+   * @returns {null} if user is not found.
+   * @returns {object} if user is found.
+   */
+  static async getUserDetails(columnName, value) {
+    const UserRecord = await user.findOne({
+      where: { [columnName]: value },
+    });
+
+    if (!UserRecord) {
```
Comment (codrex) [inline lib/modelManagers/usermodel.js]: No need for this, sequelize return null when a record is not found on a table

File: lib/utils/emailService/mailler.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+const sendMail = (to, subject, html) => {
+  const msg = {
+    to,
+    from: constants.from,
+    subject,
+    html,
+  };
+
+  if (!to && !subject && !html) {
+    return 'Parameters are expected';
```
Comment (codrex) [inline lib/utils/emailService/mailler.js]: It is better to throw an error here than returning a string describing the error.

File: lib/utils/emailService/mailler.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  if (!to || !subject || !html) {
+    return 'Incomplete Parameter(s)';
```
Comment (codrex) [inline lib/utils/emailService/mailler.js]: ☝️

File: lib/utils/emailService/mailler.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * sends E-mail.
+ * @param {string} to Receiver of mail.
+ * @param {string} subject Subject of Mail.
+ * @param {html} html Html file
+ * @returns {promise} Promise
+ */
+const sendMail = (to, subject, html) => {
+  const msg = {
```
Comment (codrex) [inline lib/utils/emailService/mailler.js:14]: This should be placed after the checks

File: lib/utils/emailService/resetMail.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    expiresIn: '30min',
+  });
+
+  // Make url string
+  const url = `${process.env.APP_BASE_URL}/api/v1/users/resetpassword/${tokenGenerated}`;
+
+  // Sends mail
+  const sendResetMail = await sendMail(verifiedEmail, 'Reset Password - Authors Haven', resetTemplate(url))
+    .then(() => 'success')
+    .catch(err => err);
```
Comment (codrex) [inline lib/utils/emailService/resetMail.js]: ` .then(() => 'success')
    .catch(err => err);
`
???

File: routes/api/forgetPassword.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+const router = require('express').Router();
+const forgetPassword = require('../../controllers/forgetPasswordController');
+
+router
+  .post('/', forgetPassword.sendResetLink)
```
Comment (codrex) [inline routes/api/forgetPassword.js:30]: add API documentation

File: test/routes/resetPassword.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+// const chai = require('chai'),
```
Comment (codrex) [inline test/routes/resetPassword.js]: remove dead code

File: .gitignore [hunk 10]
```diff
@@ -2,3 +2,4 @@ node_modules
 .nyc_output
```
Comment (iMichaelOwolabi) [inline .gitignore]: don't forget to remove this end of line

File: .gitignore [hunk 11]
```diff
@@ -2,3 +2,4 @@ node_modules
...
 .env
 .DS_Store
+.vscode
```
Comment (iMichaelOwolabi) [inline .gitignore]: kindly, take note of this end of line

Comment (Maxfurry) [reply .gitignore]: Thanks I'll correct that

File: controllers/forgetPasswordController.js (hunk 1 above)
Comment (Maxfurry) [reply controllers/forgetPasswordController.js]: Noted

File: controllers/forgetPasswordController.js (hunk 2 above)
Comment (Maxfurry) [reply controllers/forgetPasswordController.js]: Noted

File: lib/modelManagers/usermodel.js (hunk 3 above)
Comment (Maxfurry) [reply lib/modelManagers/usermodel.js]: Noted

File: lib/utils/emailService/mailler.js (hunk 4 above)
Comment (Maxfurry) [reply lib/utils/emailService/mailler.js]: Noted

File: routes/api/forgetPassword.js (hunk 8 above)
Comment (Maxfurry) [reply routes/api/forgetPassword.js:30]: Noted

File: test/routes/resetPassword.js (hunk 9 above)
Comment (Maxfurry) [reply test/routes/resetPassword.js]: Noted

File: lib/utils/emailService/resetMail.js (hunk 7 above)
Comment (Maxfurry) [reply lib/utils/emailService/resetMail.js]: The function sendMail returns a promise and so many information concerning the mail from sendgrid. So I only return error if the promise resolves to an error, if it does not I return success.

File: lib/utils/emailService/mailler.js (hunk 6 above)
Comment (Maxfurry) [reply lib/utils/emailService/mailler.js:14]: I do not understand

File: lib/utils/emailService/mailler.js (hunk 4 above)
Comment (Maxfurry) [reply lib/utils/emailService/mailler.js]: But when throwing an error isn't a message required?

File: lib/utils/emailService/mailler.js (hunk 6 above)
Comment (codrex) [reply lib/utils/emailService/mailler.js:14]: never mind

File: lib/utils/emailService/mailler.js [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    to,
+    from: FROM,
+    subject,
+    html,
+  };
+
+  // Sends mail
```
Comment (codrex) [inline lib/utils/emailService/mailler.js]: remove inline comments

File: lib/utils/emailService/resetMail.js [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+// Function that sends the reset mail with the url+token
+const resetPassword = async (verifiedEmail, uuid) => {
+  const tokenGenerated = jwtUtil.generateToken(RESET_LINK_EXPIRE, uuid);
+
+  // Make url string
+  const url = `${process.env.APP_BASE_URL}/api/v1/users/resetpassword/${tokenGenerated}`;
+
+  // Sends mail
+  const sendResetMail = await sendMail(verifiedEmail, 'Reset Password - Authors Haven', resetTemplate(url));
```
Comment (codrex) [inline lib/utils/emailService/resetMail.js]: move to constants

File: controllers/forgetPasswordController.js [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+ * @param {object} res Response Object.
+ * @returns {int} Response with status, success true or false, and message.
+ */
+  static async checkResetLink(req, res) {
+    const urlToken = req.params.token;
+    try {
+      const userId = await jwtUtil.verifyToken(urlToken).uuid;
+      res.status(200).json({
+        success: true,
+        message: 'Reset Link Valid',
```
Comment (codrex) [inline controllers/forgetPasswordController.js]: move to constants 'Reset Link Valid'

File: lib/utils/emailService/mailler.js (hunk 6 above)
Comment (Maxfurry) [reply lib/utils/emailService/mailler.js:14]: OK

File: lib/utils/emailService/mailler.js (hunk 12 above)
Comment (Maxfurry) [reply lib/utils/emailService/mailler.js]: Noted

File: controllers/forgetPasswordController.js (hunk 14 above)
Comment (Maxfurry) [reply controllers/forgetPasswordController.js]: Noted

File: lib/utils/emailService/resetMail.js (hunk 13 above)
Comment (Maxfurry) [reply lib/utils/emailService/resetMail.js]: Done

Comment (codrex) [reply lib/utils/emailService/resetMail.js]: `'Reset Password - Authors Haven'` ??

Comment (codrex) [reply lib/utils/emailService/resetMail.js]: @Maxfurry  ☝️

Comment (Maxfurry) [reply lib/utils/emailService/resetMail.js]: Noted

Comment (Maxfurry) [reply lib/utils/emailService/resetMail.js]: Fixed

**Review by codrex — APPROVED** [review]

**Pull request merged**